core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -762,55 +895,141 @@ impl<Block: BlockT> network_gossip::Validator<Block> for GossipValidator<Block>
 		-> Box<dyn FnMut(&PeerId, MessageIntent, &Block::Hash, &[u8]) -> bool + 'a>
 	{
 		let (inner, do_rebroadcast) = {
-			use parking_lot::RwLockWriteGuard;
-
-			let mut inner = self.inner.write();
+			let inner = Arc::clone(&self.inner);
 			let now = Instant::now();
-			let do_rebroadcast = if now >= inner.next_rebroadcast {
-				inner.next_rebroadcast = now + REBROADCAST_AFTER;
+			let do_rebroadcast = if now >= inner.read().next_rebroadcast {
+				(*inner.write()).next_rebroadcast = now + REBROADCAST_AFTER;
```

Comment:
Why make this update non-atomic? now it can race and `next_rebroadcast` can move backwards in time. This is a bit overzealous on the "reduce lock hold times" front because it actually breaks an invariant...

I'd recommend acquiring a read lock and then _upgrading_ it, not releasing and re-acquiring.

Reply:
Maybe there could be a tiny race, although I think it's hard to argue for it since any writing thread will first have assinged a value to `now`, then acquired the read lock, and then only acquired the write lock. So it's almost like a queue where it would be very hard for a parallel `now`to be written after the `read` that would be higher than the `now` of the current thread. 

Any way, `REBROADCAST_AFTER` is set to 5 minutes. Any potential race will be on the order or milliseconds and absolutely irrelevant. 

Finally, there is no parallelism involved in this operation, however the grandpa worker might still be waiting on the lock and happy to be able to quickly acquire and release it in between the `read` and `write` done here.


Off-course if code is written elsewhere that could hold this lock for a long time, we might be in trouble indeed. That's actually an argument for adopting this style elsewhere, and not have threads hold on to their locks to try to simulate atomic operations or transactions of some kind.  

"releasing and re-acquiring" is what parking lot is written for.

Reply:
> "releasing and re-acquiring" is what parking lot is written for.

Really tenuous claim. This is all micro-optimization anyway, and it's not worth sacrificing invariants about control flow for it.